Doc/library/tkinter.rst

```diff
@@ -91,54 +91,27 @@ Python interpreter.
+
+Tkinter implements the multithreaded model as the primary one, but it supports pumping events by hand instead of running the event loop, too.
+
+Contrary to most GUI toolkits using the multithreaded model, Tkinter calls can be made from any threads -- even worker threads. Conceptually, this can be seen as the worker thread sending an event referencing an appropriate payload, and waiting for its processing. The implementation, however, can sometimes use a shortcut here.
+
+* In threaded Tcl, an interpreter instance, when created, becomes tied to the creating OS thread. Any calls to this interpreter must come from this thread (apart from special inter-thread communication APIs). The upside is that calls to interpreters tied to different threads can run in parallel. Tkinter implements calls from outside the intrpreter thread by constructing an event with an appropriate payload, sending it to the instance's queue via the inter-thread communication APIs and waiting for result. As a consequence:
+
+    * To make any calls from outside the interpreter thread, :func:`Tk.mainloop` must be running in the interpreter thread. If it isn't, :class:`RuntimeError` is raised.
+    * A few select functions can only be run in the interpreter thread. These are the functions that implement the event loop -- :func:`Tk.mainloop`, :func:`Tk.dooneevent`, :func:`Tk.update`, :func:`Tk.update_ideltasks` -- and :func:`Tk.destroy` that terminates it halfway through.
+
+* For non-threaded Tcl, threads effectively don't exist. So, any Tkinter call is carried out in the calling thread, whatever it happens to be (see :func:`Tk.mainloop`'s entry on how it is implemented in this case). Since Tcl has a single stream of execution, all Tkinter calls are wrapped with a global lock to enforce sequential access. So, in this case, there are no restrictions on calls whatsoever, but only one call, to any interpreter, can be active at a time.
+
+The last thing to note is that Tcl event queus are not per-interpreter but rather per-thread. So, a running event loop will process events not only for its own interpreter, but also for any others that share the same thread. This is transparent for the code though because an event handler is invoked within the context of the correct interpreter (and in the correct Python lexical context if the handler has a Python payload). There's also no harm in trying to run an event loop for two interpreters that may happen to share a queue: in threaded Tcl, such a clash is flat-out impossible because they would have to both run in the same OS thread, and in non-threaded Tcl, they would take turns processing events.
```

Comment:
I take it this is only relevant if you have more than one Tkinter() object instantiated at the same time. (Why would anyone do that? Has there even been a promise made that you could have two separate Tkinter instances in a process at the same time? )

Reply:
Tkinter doesn't have any code that intentionally limits the ability to operate multiple interpreters. So, it is designed to support this scenario. So this is a part of its public interface and as such, the domain of a reference documentation.

A user may wish to run `Tcl()`s (interpreters without Tk loaded) in them or something to run some Tcl program, or for some satellite UIs. As a doc writer, that's none of my business really. I've got enough experience to know that users can and will invent ways to use your code beyond your wildest imagination.

Reply:
I'm not against dealing with every possible contingency in the reference docs, though I am against people having to wade through it for what is in practice a very fringe case.

We both know very well that just because there's no code to limit something from happening it doesn't mean it's designed to support it. 

From a Tcl/Tk perspective, running multiple instances of Tk has always been considered a bad idea ("doctor it hurts when I do that... don't do that"). It's grudgingly been accepted that it works when done in a separate thread, though still nobody thinks it's a good idea. Tkinter should be following that lead in at least recommending not to do it. Running multiple Tcl interpreters, threads or not, however, is definitely supported and encouraged.

Having said all that, it would still be rare for a Python programmer to run more than one Tcl interpreter (i.e. it would just be the one created by Tkinter). So let's deal with the two cases separately in the Tkinter docs.

First part would be for if you're running only a single Tcl/Tk interpreter, you need to worry about X, Y, Z. (Main concern would be not running things that prevent Tcl_DoOneEvent from being called.)

Second part (which we can assume will be for expert users, and can therefore use as terse terminology as you'd like) would be for if you have more than one Tcl interpreter. In that case, the stuff you've said (with some corrections) would apply. I'd use pointers to Tcl/Tk docs liberally (e.g. https://www.tcl.tk/man/tcl8.6/TclLib/Notifier.htm) rather than explaining things in the Tkinter docs where possible. 

Fencing off the second part like this, so most people know not to look at it, would work for me. How about you?

Reply:
I never mentioned _Tk_ anywhere in the paragraph. Any Tcl interpreter instance must process its event queue to be able to handle asynchronous events. Tk is irrelevant here, and unless you can point me to an official doc saying that Tk imposes additional limitations here over what plain Tcl machinery allows, I don't see a need to bring it up at all.

Now, the fact that Tcl event queue is shared is one of the critical things to understand Tkinter's behavior. Note that I describe the standard GUI execution model, then bring up all cases where Tkinter significantly differs from it. This is one of them.

And it's definitely related to threading model, so it belongs in this section rather than somewhere else.

Subjectively splitting it into "basic" and "advanced" is against ref doc design and is actively detrimental to its value. By its purpose, a ref doc is intended to fill in the gaps rather than teach how to use the product. So, the most critical thing is to be able to quickly find the required topic, and to be sure that you've seen everything on it. A subjective split breaks both these promises.

Reply:
I even wanted to add a warn about nested event loops here, too, but ultimately decided against it 'cuz it's a bad usage pattern rather than API restriction.

Reply:
That said, "There's also no harm... etc" smells like sidetracking: e.g. that the loops will "take turns" is not really an official promise.

Reply:
We seem to be at a bit of an impasse here. I suggest we invite some other people to provide feedback or suggestions, whether on process or substantive content. Based on my limited knowledge I'd suggest @serhiy-storchaka, @terryjreedy, and @zware. 

Quick version of main point of contention is contained a few comments up, in mine starting "I'm not against..." (particularly last four paragraphs) and @native-api's following one starting "I never mentioned...". Essentially who needs to know what level of detail and how that looks in these docs. Thanks.

Reply:
When making edits from Github UI, you can choose to make a separate PR suggesting them instead of committing them to this one.

I suggest you do that. When you'll be actually writing stuff and thinking about what it should look like, all the downsides I'm talking about should become more visible.